ferretdb/ferretdb.go

```diff
@@ -0,0 +1,149 @@
+)
+
+// Config contains a PostgreSQL backend connection string.
+type Config struct {
+	PostgreSQLConnectionString string
+}
+
+// FerretDB proxy.
+type FerretDB struct {
+	config   Config
+	mongoURL string
+}
+
+// New returns a new FerretDB.
+func New(conf Config) FerretDB {
+	return FerretDB{
+		config:   conf,
+		mongoURL: transformConnStringPgToMongo(conf.PostgreSQLConnectionString),
+	}
+}
+
+// Run runs the FerretDB proxy as a library with
+// * error level logging
+// * monitoring disabled
+// * handler PostgreSQL.
+func (fdb *FerretDB) Run(ctx context.Context, conf Config) (string, error) {
+	listenAddr := "127.0.0.1:27017"
+	proxyAddr := "127.0.0.1:37017"
+	debugAddr := "127.0.0.1:8088"
+	mode := clientconn.NormalMode
+	handler := "pg"
+	testConnTimeout := time.Duration(0)
+
+	_, ok := register.HandlerFunc["pg"]
+	if !ok {
+		panic("no pg handler registered")
+	}
+
+	logging.Setup(zapcore.ErrorLevel)
+	logger := zap.L()
+
+	info := version.Get()
+
+	startFields := []zap.Field{
+		zap.String("version", info.Version),
+		zap.String("commit", info.Commit),
+		zap.String("branch", info.Branch),
+		zap.Bool("dirty", info.Dirty),
+	}
+	for _, k := range info.BuildEnvironment.Keys() {
+		v := must.NotFail(info.BuildEnvironment.Get(k))
+		startFields = append(startFields, zap.Any(k, v))
+	}
+
+	go debug.RunHandler(ctx, debugAddr, logger.Named("debug"))
+
+	newHandler := register.HandlerFunc[handler]
+	if newHandler == nil {
+		logger.Sugar().Fatalf("Unknown backend handler %q.", handler)
+	}
+	h, err := newHandler(&register.NewHandlerOpts{
+		PostgreSQLConnectionString: conf.PostgreSQLConnectionString,
+		Ctx:                        ctx,
+		Logger:                     logger,
+	})
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+	defer h.Close()
+
+	l := clientconn.NewListener(&clientconn.NewListenerOpts{
+		ListenAddr:      listenAddr,
+		ProxyAddr:       proxyAddr,
+		Mode:            clientconn.Mode(mode),
+		Handler:         h,
+		Logger:          logger,
+		TestConnTimeout: testConnTimeout,
+	})
+
+	err = l.Run(ctx)
+	if err == nil || err == context.Canceled {
+		logger.Info("Listener stopped")
+	} else {
+		logger.Error("Listener stopped", zap.Error(err))
+	}
+	return fdb.mongoURL, nil
+}
+
+// transformConnStringPgToMongo parses postgresql connection string and
+// returns a corresponded MongoDB connection string for the driver.
+// I.e. transforms "postgres://postgres@127.0.0.1:5432/ferretdb" into mongodb://127.0.0.1:27017.
+func transformConnStringPgToMongo(connectionURL string) string {
+	u, err := url.Parse(connectionURL)
+	if err != nil {
+		panic(err)
+	}
+
+	// change scheme
+	if u.Scheme != "postgres" {
+		panic("connection url: postgres scheme required")
+	}
+	u.Scheme = "mongodb"
+
+	// MongoDB collections corresponds to scheme in PostgreSQL
+	u.Path = ""
+
+	// set the port to the FerreDB port
+	host, _, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		panic(err)
+	}
+	port := "27017"
+	u.Host = host + ":" + port
```

Comment:
Embedded FerretDB always listens on 127.0.0.1:27017 (that is `listenAddr` above). PostgreSQL does not have to be on the same host. So using PostgreSQL's host is not correct.

The whole function really should just return a fixed string `mongodb://127.0.0.1:27017`. And then we probably don't need a separate function.

Reply:
> 127.0.0.1:27017

That's for now.
Why don't you think people won't create a PostgreSQL on some server(s) other than FerretDB? 
They can create a proxy and pass it to their own PostgreSQL creating Atlas replacement by this.

Reply:
> Embedded FerretDB always listens on 127.0.0.1:27017 (that is `listenAddr` above).

And the Port could be changed also - i.e. for security reasons.

Reply:
> Why don't you think people won't create a PostgreSQL on some server(s) other than FerretDB?

I think people can do that. And if we pass `somehost:5432` to this function we will get `somehost:27017` in MongoDB URI – while we should get `127.0.0.1:27017`, because embedded FerretDB runs on this host, not on `somehost`

Reply:
So if I understand the `transformConnStringPgtoMongo` after this change just checks if connectionURL `scheme` is postgres?